subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputPropertiesIntegrationTest.groovy

```diff
@@ -417,31 +414,129 @@ apply from:'scriptPlugin.gradle'
                 FooType foo
+            }
+
+            task createFile(type: MyTask) {
+                ext.outputFile = file('output.txt')
+                outputs.file(outputFile)
+
+                foo = FooType.FOO
+
+                doLast {
+                    outputFile << foo
+                }
+            }
+"""
+
+        when:
+        succeeds 'createFile'
+
+        then:
+        executedTasks == [':createFile']
+        skippedTasks.empty
+
+        when:
+        succeeds 'createFile'
 
-                MyTask() {
-                    outputs.upToDateWhen {
-                        false
+        then:
+        executedTasks == [':createFile']
+        skippedTasks == [':createFile'] as Set
+    }
+
+    @Issue("gradle/gradle#919")
+    def "can use a custom Serializable type from init script as input property in this taskType: #taskType"() {
+        def initScript = file("init.gradle")
+
+        given:
+        initScript << """
+            enum FooType { FOO }
+
+            rootProject {
+                afterEvaluate {
+                    tasks.createFile {
+                        ext.fooType = FooType.FOO
+                        inputs.property('fooType', fooType)
                     }
                 }
             }
+"""
 
-            task neverUpToDate(type: MyTask) {
-                foo = FooType.FOO
+        buildFile << """
+            import org.gradle.api.internal.changedetection.state.InMemoryTaskArtifactCache
+
+            println "Flushing InMemoryTaskArtifactCache"
+            gradle.taskGraph.whenReady {
+                gradle.services.get(InMemoryTaskArtifactCache).invalidateAll()
+            }
+
+            class MyTask extends DefaultTask {}
+
+            task createFile(type: $taskType) {
+                ext.outputFile = file('output.txt')
+                outputs.file(outputFile)
+
+                doLast {
+                    outputFile << "some data"
+                }
             }
 """
 
         when:
-        succeeds 'neverUpToDate'
+        args "-I", initScript.absolutePath
+        succeeds 'createFile'
 
         then:
-        executedTasks == [':neverUpToDate']
+        executedTasks == [':createFile']
         skippedTasks.empty
 
         when:
-        succeeds 'neverUpToDate'
+        args "-I", initScript.absolutePath
+        succeeds 'createFile'
+
+        then:
+        executedTasks == [':createFile']
+        skippedTasks == [':createFile'] as Set
+
+        where:
+        taskType << ['DefaultTask', 'MyTask']
+    }
+
+    @Issue("gradle/gradle#919")
+    def "show deprecation warning when using a custom Serializable type with diverging behavior between equals and object hash"() {
+        given:
+        buildFile << """
+            class FooType implements Serializable {
+                @Override
+                boolean equals(Object obj) {
+                    return false;
+                }
+            }
+
+            task createFile {
```

Comment:
I'm not sure what is the advantage of what you are suggesting. Although the type has the same name, they have a different purpose. The `FooType` in this test exercise a custom `equals` implementation which is irrelevant and possibly erroneous for the tests above. The same goes for the task type and usage. I do want to configure a task that exercises the up-to-date check but I don't necessary want a custom type as the task type class loader did have an effect on the `ClassNotFoundException` issue we are trying to solve. I think it would be wise to leave it as is. What do you think?